DataFormats/Detectors/EMCAL/include/DataFormatsEMCAL/Cluster.h

```diff
@@ -0,0 +1,89 @@
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#ifndef ALICEO2_EMCAL_CLUSTER_H_
+#define ALICEO2_EMCAL_CLUSTER_H_
+
+#include <array>
+#include <exception>
+#include <iosfwd>
+#include <string>
+#include <vector>
+#include "CommonDataFormat/TimeStamp.h" 
+#include "MathUtils/Cartesian3D.h"
+
+namespace o2
+{
+
+namespace emcal
+{
+
+/// \class Cluster
+/// \brief EMCAL Cluster
+///
+class Cluster : public o2::dataformats::TimeStamp<Double_t>, public Point3D<Double_t>
+{
+public:
+    class InvalidPositionException : public std::exception{
+    public:
+        InvalidPositionException() = default;
+        InvalidPositionException(const std::array<double,3> &pos) : std::exception(), mPosition(pos) {}
+        ~InvalidPositionException() noexcept override = default;
+
+        const char *what() const noexcept override {
+            return Form("Invalid position: %f, %f, %f", mPosition[0], mPosition[1], mPosition[2]);
+        }
+    private:
+        std::array<double, 3>           mPosition;
+        std::string                     mMessage;
+    };
+
+    Cluster() = default;
+    Cluster(Point3D<double> pos, Double_t time, Double_t energy);
+    Cluster(Double_t x, Double_t y, Double_t z, Double_t time, Double_t energy);
+    ~Cluster() noexcept = default;
+
+    Double_t getM02() const { return mM02; }
+    Double_t getM20() const { return mM20; }
+    Double_t getDispersion() const { return mDispersion; }
+    Double_t getDistanceToBadCell() const { return mDistanceToBadCell; }
+    Bool_t isExotic() const { return mIsExotic; }
+    Int_t getNCells() const { return mCellIndices.size(); }
+    std::vector<unsigned short> getCellIndices() const { return mCellIndices; }
+
+    void setIsExotic(Bool_t exotic) { mIsExotic = exotic; }
+    void setM02(Double_t val) { mM02 = val; }
+    void setM20(Double_t val) { mM20 = val; }
+    void setDispersion(Double_t val) { mDispersion = val;}
+    void setDistanceToBadCell(Double_t distance) {mDistanceToBadCell = distance; }
+    void setCellIndices(const std::vector<unsigned short> cells) { mCellIndices = cells; }
+
+    Vector3D<double> getMomentum(const Point3D<double> *vertex) const;
+
+    void PrintStream(std::ostream &stream) const;
+    
+private:
+    Double_t        mEnergy;            ///< Cluster energy
+    Double_t        mDispersion;        ///< Cluster dispersion
+    Double_t        mDistanceToBadCell; ///< Distance to nearest bad cell
+    Double_t        mM02;               ///< M02 parameter (2-nd moment along the main eigen axis.)
+    Double_t        mM20;               ///< M20 parameter (2-nd moment along the second eigen axis.)
+    Bool_t          mIsExotic;          ///< Mark cluster as exotic cluster
+    std::vector<unsigned short> mCellIndices; ///< Cells contributing to a cluser
```

Comment:
At the moment it is not so important, but it is preferable not to have dynamic members in copiously produced classes. The more economical approach is to have separately a vector of cell indices used by clusters and to refer from the cluster to the list of its cells using either ``RangeReference`` or ``RangeRefComp`` from https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/common/include/CommonDataFormat/RangeReference.h, see e.g. https://github.com/AliceO2Group/AliceO2/blob/dev/DataFormats/Detectors/ITSMFT/ITS/include/DataFormatsITS/TrackITS.h and its usage.